File: 212.java

class Trie {
    public Trie[] next = new Trie[26];
    public String word = null;
    public Trie() {}
}


class Solution {

    private void dfs(List<String> rightWord, char[][] board, boolean[][] visited, Trie node, int x, int y) {
        
        if (x == -1 || y == -1 || x == board.length || y == board[0].length)
            return;
        if (visited[x][y]) return;
        int idx = board[x][y] - 'a';
        if (node.next[idx] == null) return;
        node = node.next[idx];
        if (node.word != null) {
            rightWord.add(node.word);
            node.word = null;
        }
        visited[x][y] = true;
        dfs(rightWord, board, visited, node, x, y - 1);
        dfs(rightWord, board, visited, node, x, y + 1);
        dfs(rightWord, board, visited, node, x - 1, y);
        dfs(rightWord, board, visited, node, x + 1, y);
        visited[x][y] = false;
    }

    public List<String> findWords(char[][] board, String[] words) {
        Trie root = new Trie();
        for (String word : words) {
            Trie node = root;
            for (int i = 0; i < word.length(); ++i) {
                int idx = word.charAt(i) - 'a';
                if (null == node.next[idx]) {
                    node.next[idx] = new Trie();
                }
                node = node.next[idx];
            }
            node.word = word;
        }
        List<String> rightWord = new LinkedList<>();
        for (int i = 0; i < board.length; ++i) {
            for (int j = 0; j < board[0].length; ++j) {
                dfs(rightWord, board, new boolean[board.length][board[0].length], root, i, j);
            }
        }
        return rightWord;
    }
}

File: 212.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

typedef struct _Trie_node {
    struct _Trie_node * next[26];
    char * endWord;
} Trie_node;

void dfs(char** board, unsigned char * visited, int n, int m, int x, int y, Trie_node * node, int * returnSize, char ** returnStr) {
    if (x == -1 || y == -1 || x == n || y == m) return;
    int arridx = m * x + y;
    if (visited[arridx]) return;
    int idx = board[x][y] - 'a';
    node = node->next[idx];
    if (!node) return;
    if (node->endWord) {
        returnStr[(*returnSize) ++] = node->endWord;
        node->endWord = NULL;
    }
    visited[arridx] = 1;
    dfs(board, visited, n, m, x - 1, y, node, returnSize, returnStr);
    dfs(board, visited, n, m, x + 1, y, node, returnSize, returnStr);
    dfs(board, visited, n, m, x, y - 1, node, returnSize, returnStr);
    dfs(board, visited, n, m, x, y + 1, node, returnSize, returnStr);
    visited[arridx] = 0;
}

char** findWords(char** board, int boardSize, int* boardColSize, char** words, int wordsSize, int* returnSize) {
    int n = boardSize, m = boardColSize[0];
    unsigned char* visited = (unsigned char *)malloc(n * m * sizeof(unsigned char));
    Trie_node * root = (Trie_node *)calloc(1, sizeof(Trie_node));
    for (int i = 0; i < wordsSize; ++i) {
        Trie_node * node = root;
        char * pointer = words[i];
        while (*pointer != '\0') {
            int idx = *pointer - 'a';
            if (node->next[idx] == NULL) {
                node->next[idx] = (Trie_node *)calloc(1, sizeof(Trie_node));
            }
            node = node->next[idx];
            pointer ++;
        }
        node->endWord = words[i];
    }
    *returnSize = 0;
    char ** returnStr = (char **)calloc(wordsSize, sizeof(char *));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            memset(visited, 0, n * m * sizeof(unsigned char));
            dfs(board, visited, n, m, i, j, root, returnSize, returnStr);
        }
    }
    return returnStr;
}